File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/ScreenProtection.kt
package dev.idenqa.sdk

import android.view.Window
import android.view.WindowManager

/**
 * Sensitive-screen protection port.
 *
 * The port never fails a capture journey when the platform cannot protect a
 * surface: `isAvailable` reports the honest state and the journey degrades to
 * guidance. `isCaptured` reports recording/mirroring state when the platform
 * exposes it; Android does not provide a reliable capture flag, so the default
 * implementation reports false rather than guessing.
 */
interface ScreenProtection {
    val isAvailable: Boolean
    suspend fun applyProtection()
    suspend fun removeProtection()
    suspend fun isCaptured(): Boolean
}

/** Graceful default used when a platform protection surface is unavailable. */
object NoopScreenProtection : ScreenProtection {
    override val isAvailable: Boolean get() = false
    override suspend fun applyProtection() {}
    override suspend fun removeProtection() {}
    override suspend fun isCaptured(): Boolean = false
}

/**
 * Applies `FLAG_SECURE` to one host window while a capture preview is visible.
 * The flag prevents screenshots, screen recording, and non-secure displays.
 * Where an OEM rejects the flag the adapter degrades to `isAvailable = false`
 * rather than failing the journey.
 */
class AndroidWindowScreenProtection(private val window: Window) : ScreenProtection {
    private var applied = false

    private var rejected = false

    override val isAvailable: Boolean get() = !rejected

    override suspend fun applyProtection() {
        if (rejected) return
        try {
            window.addFlags(WindowManager.LayoutParams.FLAG_SECURE)
            applied = true
        } catch (_: Exception) {
            rejected = true
            applied = false
        }
    }

    override suspend fun removeProtection() {
        if (!applied) return
        try {
            window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
        } catch (_: Exception) {
            rejected = true
        } finally {
            applied = false
        }
    }

    override suspend fun isCaptured(): Boolean = false
}

File: sdk/swift/Sources/Idenqa/ScreenProtection.swift
import Foundation

/// Sensitive-screen protection port.
///
/// The port never fails a capture journey when the platform cannot protect a
/// surface: `isAvailable` reports the honest state and the journey degrades to
/// guidance. `isCaptured` reports recording/mirroring state when the platform
/// exposes it.
public protocol CaptureScreenProtection: Sendable {
    var isAvailable: Bool { get }
    func applyProtection() async
    func removeProtection() async
    func isCaptured() async -> Bool
}

/// Graceful default used when a platform protection surface is unavailable.
public struct NoopCaptureScreenProtection: CaptureScreenProtection {
    public init() {}
    public var isAvailable: Bool { false }
    public func applyProtection() async {}
    public func removeProtection() async {}
    public func isCaptured() async -> Bool { false }
}

#if os(iOS) && canImport(UIKit)
import UIKit

/// iOS sensitive-screen protection.
///
/// `isCaptured` reflects `UIScreen.isCaptured` (screen recording, mirroring, or
/// AirPlay capture). Apply protection before presenting a capture preview and
/// remove it when the preview leaves the screen.
///
/// `CaptureSensitiveView` masks its contents while the screen is captured.
@MainActor
public final class IOSSensitiveScreenProtection: CaptureScreenProtection {
    public nonisolated let isAvailable = true
    private var protected = false

    public init() {}

    public func applyProtection() async {
        protected = true
    }

    public func removeProtection() async {
        protected = false
    }

    public func isCaptured() async -> Bool {
        UIScreen.main.isCaptured
    }

    public var isProtectionApplied: Bool { protected }
}

/// A view that masks its content while the screen is being captured or recorded.
@MainActor
public final class CaptureSensitiveView: UIView {
    private let maskLayer = UIView()

    public var maskColor: UIColor = .systemBackground {
        didSet { maskLayer.backgroundColor = maskColor }
    }

    public override init(frame: CGRect) {
        super.init(frame: frame)
        configure()
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        configure()
    }

    private func configure() {
        maskLayer.backgroundColor = maskColor
        maskLayer.isHidden = true
        maskLayer.isUserInteractionEnabled = false
        maskLayer.translatesAutoresizingMaskIntoConstraints = false
        addSubview(maskLayer)
        NSLayoutConstraint.activate([
            maskLayer.leadingAnchor.constraint(equalTo: leadingAnchor),
            maskLayer.trailingAnchor.constraint(equalTo: trailingAnchor),
            maskLayer.topAnchor.constraint(equalTo: topAnchor),
            maskLayer.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(captureStateChanged),
            name: UIScreen.capturedDidChangeNotification,
            object: nil
        )
        captureStateChanged()
    }

    @objc private func captureStateChanged() {
        maskLayer.isHidden = !UIScreen.main.isCaptured
    }
}
#endif
